compiler/rustc_driver/src/lib.rs

```diff
@@ -1202,7 +1202,10 @@ pub fn report_ice(info: &panic::PanicInfo<'_>, bug_report_url: &str) {
         handler.note_without_error(&note);
     }
 
-    TyCtxt::try_print_query_stack(&handler, Some(2));
+    // If backtraces are enabled, also print the query stack
+    let backtrace = env::var_os("RUST_BACKTRACE").map(|x| &x != "0").unwrap_or(false);
+
+    TyCtxt::try_print_query_stack(&handler, Some(2), Some(backtrace));
```

Comment:
you could also keep a single (`num_frames`) argument and set it to `None` if `backtrace` and to `Some(2)` otherwise and then check `num_frames == Some(i)` in the loop.